### PR TITLE
IGNITE-18322 Define scan contract SortedIndexStorage

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-18322

### CODE REVIEW

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 1]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+
+        assertEquals(0, serializer.deserializeColumns(nextRow)[0]);
+        assertEquals(rowId0, nextRow.rowId());
+
+        assertFalse(scan.hasNext());
+        assertThrows(NoSuchElementException.class, scan::next);
+
+        put(indexStorage, serializer.serializeRow(new Object[]{0}, rowId0));
+
+        assertFalse(scan.hasNext());
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:865]: You know what, at each step it would be cool ta have a comment that shows the state of the index and the state of the cursor. Like `0 1 2 | 3 4` or something, otherwise it might be complicated to remember everything.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 2]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+
+        assertEquals(1, serializer.deserializeColumns(nextRow)[0]);
+        assertEquals(rowId0, nextRow.rowId());
+
+        remove(indexStorage, serializer.serializeRow(new Object[]{2}, rowId1));
+
+        assertFalse(scan.hasNext());
+        assertThrows(NoSuchElementException.class, scan::next);
+    }
+
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:982]: We need more tests. What if current position is removed but the new entry is inserted into "the same" place? Or n the place strictly before, for example. Combinations of insertions and deletions should be checked, at least some of them.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java [hunk 3]
```diff
@@ -45,7 +44,9 @@
  * Test implementation of MV sorted index storage.
  */
 public class TestSortedIndexStorage implements SortedIndexStorage {
-    private final ConcurrentNavigableMap<ByteBuffer, Set<RowId>> index;
+    private static final Object NULL = new Object();
+
+    private final ConcurrentNavigableMap<ByteBuffer, NavigableMap<RowId, Object>> index;
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: I guess `NavigableSet` should be used instead of the map. What do you store in those values?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java [hunk 4]
```diff
@@ -167,24 +167,71 @@ private Cursor<ByteBuffer> createScanCursor(byte @Nullable [] lowerBound, byte @
...
+                    } catch (RocksDBException e) {
+                        throw new StorageException("Error refreshing an iterator", e);
+                    }
+
+                    if (key == null) {
+                        it.seek(lowerBound == null ? partitionStorage.partitionStartPrefix() : lowerBound);
+                    } else {
+                        it.seekForPrev(key);
+
+                        it.next();
```
Comment (ibessonov) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:219]: Technically, iterator may not be valid in this case and calling `next` is unsafe. You should add a test where index has a single value and it's deleted while iterator is "standing" on it.
It'll most likely fail.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java [hunk 5]
```diff
@@ -208,4 +184,91 @@ private void checkClosed() {
...
+                if (indexMapEntry == null) {
+                    indexMapEntry = indexMap.firstEntry();
+                }
+
+                if (rowId == null) {
+                    if (indexMapEntry != null) {
+                        rowId = getRowId(indexMapEntry.getValue().firstEntry());
+                    }
+                } else {
+                    Entry<RowId, Object> rowIdEntry = indexMapEntry.getValue().higherEntry(rowId);
```
Comment (rpuch) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:246]: Should this be named `nextRowIdEntry`?

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (rpuch) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: It looks like this can still be `ConcurrentNavigableMap<ByteBuffer, NavigableSet<RowId>>` which is simpler to use and it causes less confusion (because something that is used as a set is called Set, not Map)

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java [hunk 6]
```diff
@@ -272,4 +271,97 @@ public R next() {
...
+                return toIndexRowImpl(treeRow);
+            } catch (IgniteInternalCheckedException e) {
+                throw new StorageException("Error while advancing the cursor", e);
+            } finally {
+                closeBusyLock.leaveBusy();
+            }
+        }
+
+        private void advanceIfNeeded() throws IgniteInternalCheckedException {
+            if (hasNext == null) {
```
Comment (rpuch) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java:341]: Early return pattern could be used to decrease indentation level for the principal logic of the method

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 7]
```diff
@@ -1572,6 +1572,34 @@ public final T findOne(L row) throws IgniteInternalCheckedException {
         return findOne(row, null, null);
     }
 
+    /**
+     * Searches for the next row from the passed from the arguments.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1576]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 8]
```diff
@@ -1572,6 +1572,34 @@ public final T findOne(L row) throws IgniteInternalCheckedException {
...
+     *
+     * @param row Row.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1578]: Yes, we can use this name.

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java [hunk 9]
```diff
@@ -167,24 +167,71 @@ private Cursor<ByteBuffer> createScanCursor(byte @Nullable [] lowerBound, byte @
...
+            @Nullable
+            private Boolean hasNext;
 
-        return new RocksIteratorAdapter<>(it) {
-            @Override
-            protected ByteBuffer decodeEntry(byte[] key, byte[] value) {
-                return ByteBuffer.wrap(key).order(ORDER);
-            }
+            @Nullable
+            private byte[] key;
```
Comment (rpuch) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:175]:
```suggestion
            private byte @Nullable [] key;
```

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java [hunk 10]
```diff
@@ -167,24 +167,71 @@ private Cursor<ByteBuffer> createScanCursor(byte @Nullable [] lowerBound, byte @
...
 
             @Override
             public void close() {
-                super.close();
+                it.close();
```
Comment (rpuch) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:179]: Let's include `it` into the call to `closeAll()`

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 11]
```diff
@@ -6222,6 +6250,72 @@ public void close() {
...
+
+            int cnt = io.getCount(pageAddr);
+
+            if (cnt == 0) {
+                // Empty tree.
+                assert io.getForward(pageAddr, partId) == 0L;
+            } else {
+                assert io.isLeaf() : io;
+                assert cnt > 0 : cnt;
+                assert idx >= 0 || idx == -1 : idx;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:6286]: You're right, I looked at the code in neighboring classes and there was this.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 12]
```diff
@@ -6222,6 +6250,72 @@ public void close() {
...
+                assert cnt >= idx : "cnt=" + cnt + ", idx=" + idx;
+
+                checkDestroyed();
+
+                if (idx == -1) {
+                    idx = findNextRowIdx(pageAddr, io, cnt);
+                }
+
+                if (cnt - idx > 0) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:6295]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 13]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+            put(indexStorage, serializer.serializeRow(new Object[]{i}, new RowId(TEST_PARTITION)));
+        }
+
+        // Checking without borders.
+        assertThat(
+                scan(indexStorage, null, null, 0).stream().map(objects -> objects[0]).collect(toList()),
+                contains(0, 1, 2, 3, 4)
+        );
+
+        // Let's check without borders.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:611]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 14]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+    void testScanContractForEmptyIndex() {
+        SortedIndexDefinition indexDefinition = SchemaBuilders.sortedIndex("TEST_IDX")
+                .addIndexColumn(ColumnType.INT32.typeSpec().name()).asc().done()
+                .build();
+
+        SortedIndexStorage indexStorage = createIndexStorage(indexDefinition);
+
+        BinaryTupleRowSerializer serializer = new BinaryTupleRowSerializer(indexStorage.indexDescriptor());
+
+        Cursor<IndexRow> scanBeforeHashNext = indexStorage.scan(null, null, 0);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:757]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1576]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:6295]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 15]
```diff
@@ -6222,6 +6250,72 @@ public void close() {
...
+                    nextRow = getRow(io, pageAddr, idx);
+                }
+            }
+
+            return true;
+        }
+
+        int findNextRowIdx(long pageAddr, BplusIo<L> io, int cnt) throws IgniteInternalCheckedException {
+            // Compare with the first row on the page.
+            int cmp = compare(0, io, pageAddr, 0, row);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:6305]: We will never get into this code because the `idx` is always >=0, I deleted this code.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 16]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+
+        assertThat(
+                indexStorage.scan(
+                                serializer.serializeRowPrefix(0),
+                                serializer.serializeRowPrefix(4),
+                                (GREATER_OR_EQUAL | LESS)
+                        ).stream()
+                        .map(indexRow -> serializer.deserializeColumns(indexRow)[0])
+                        .collect(toList()),
+                contains(0, 1, 2, 3)
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:631]: In `org.hamcrest.Matchers#contains(E...)`, it is checked for equality by the number of elements, I looked at the documentation and tried it with my hands.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 17]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
         }
     }
 
+    /**
+     * Checks simple scenarios for a scanning cursor.
+     */
+    @Test
+    void testScanSimple() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:592]: In general, it can be done, but since it is now clearer for me.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 18]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+        Cursor<IndexRow> scanWithoutHasNext = indexStorage.scan(null, null, 0);
+
+        put(indexStorage, serializer.serializeRow(new Object[]{0}, new RowId(TEST_PARTITION)));
+
+        assertTrue(scanBeforeHashNext.hasNext());
+        assertFalse(scanAfterHasNext.hasNext());
+
+        assertEquals(0, serializer.deserializeColumns(scanBeforeHashNext.next())[0]);
+
+        assertThrows(NoSuchElementException.class, scanAfterHasNext::next);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:772]: Method `java.util.Iterator#hasNext` and `java.util.Iterator#next` should be consistent, I don't think it needs to be described in the documentation.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 19]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+
+        BinaryTupleRowSerializer serializer = new BinaryTupleRowSerializer(indexStorage.indexDescriptor());
+
+        for (int i = 0; i < 5; i++) {
+            put(indexStorage, serializer.serializeRow(new Object[]{i}, new RowId(TEST_PARTITION)));
+        }
+
+        // Checking without borders.
+        assertThat(
+                scan(indexStorage, null, null, 0).stream().map(objects -> objects[0]).collect(toList()),
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:607]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 20]
```diff
@@ -579,6 +585,401 @@ void testNullValues(ColumnDefinition columnDefinition) throws Exception {
...
+                        ).stream()
+                        .map(indexRow -> serializer.deserializeColumns(indexRow)[0])
+                        .collect(toList()),
+                contains(0, 1, 2)
+        );
+    }
+
+    @Test
+    void testScanContractForEmptyIndex() {
+        SortedIndexDefinition indexDefinition = SchemaBuilders.sortedIndex("TEST_IDX")
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:749]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:757]: Divided into three methods, I think it should become clearer.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java (hunk 1 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:865]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java (hunk 2 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:982]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: Main problem in `java.util.SortedSet#first` if the set is empty it will throw an exception.

Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: Main problem in `java.util.SortedSet#first` if the set is empty it will throw an exception.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 5 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:246]: Fix it

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java (hunk 6 above)
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/sorted/PageMemorySortedIndexStorage.java:341]: Fix it

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java (hunk 9 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:175]: Fix it

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java (hunk 10 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:179]: Why?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java (hunk 4 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:219]: Tried to test in `org.apache.ignite.internal.storage.index.AbstractSortedIndexStorageTest#testScanContractRemoveCachedRow`

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (rpuch) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: Well, I changed the code to migrate it from maps to sets, and tests passed. This is probably because the set/map is never empty (if we remove the last element of it, we remove the set/map from the `index`). So the fact that `SortedSet#first()` throws for an empty set does not seem to be a problem.

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java (hunk 10 above)
Comment (rpuch) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:178]: To make sure that we close as many resources as possible even if an exception happens when closing one of them. This is probably the reason `closeAll()` exists :)

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: If we look at `TestSortedIndexStorage#remove`, we can get into trouble if, after getting "RowId"s, in parallel delete first element (become empty RowIds) it and then try to take the first element.

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java (hunk 10 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:178]: Fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (rpuch) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: Yes, you are right. But we can just wrap the call to `set.first()` with `try`/`catch NoSuchElementException`. This needs to be done in just 2 places, and we'll have a set represented by a Set, not by a Map. Still less confusion.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 21]
```diff
@@ -1573,17 +1573,18 @@ public final T findOne(L row) throws IgniteInternalCheckedException {
     }
 
     /**
-     * Searches for the next row from the passed from the arguments.
+     * Searches for the lowerBound that (strictly or loosely, depending on {@code includeRow}) follows the lowerBound passed as an
```
Comment (rpuch) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1576]:
```suggestion
     * Searches for the row that (strictly or loosely, depending on {@code includeRow}) follows the lowerBound passed as an
```

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 22]
```diff
@@ -1573,17 +1573,18 @@ public final T findOne(L row) throws IgniteInternalCheckedException {
...
+     * argument.
      *
-     * @param row Row.
-     * @param includeRow {@code True} if you include the passed row in the result.
-     * @return Next row.
+     * @param lowerBound Lower bound..
+     * @param includeRow {@code True} if you include the passed lowerBound in the result.
+     * @return Next lowerBound.
```
Comment (rpuch) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1581]:
```suggestion
     * @param includeRow {@code True} if you include the passed row in the result.
     * @return Next row.
```

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 23]
```diff
@@ -1592,9 +1593,9 @@ public final T findOne(L row) throws IgniteInternalCheckedException {
         } catch (CorruptedDataStructureException e) {
             throw e;
         } catch (IgniteInternalCheckedException e) {
-            throw new IgniteInternalCheckedException("Runtime failure on lookup next row: " + row, e);
+            throw new IgniteInternalCheckedException("Runtime failure on lookup next lowerBound: " + lowerBound, e);
```
Comment (rpuch) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1596]:
```suggestion
            throw new IgniteInternalCheckedException("Runtime failure on lookup next row: " + lowerBound, e);
```

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 24]
```diff
@@ -1592,9 +1593,9 @@ public final T findOne(L row) throws IgniteInternalCheckedException {
...
         } catch (RuntimeException | AssertionError e) {
-            throw corruptedTreeException("Runtime failure on lookup next row: " + row, e, grpId, g.pageId);
+            throw corruptedTreeException("Runtime failure on lookup next lowerBound: " + lowerBound, e, grpId, g.pageId);
```
Comment (rpuch) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1598]:
```suggestion
            throw corruptedTreeException("Runtime failure on lookup next row: " + lowerBound, e, grpId, g.pageId);
```

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: It's not good to write code based on exceptions, it's better to leave the map.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java (hunk 23 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1596]: fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java (hunk 24 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1598]: fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java (hunk 21 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:1576]: fix it

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (rpuch) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: It's ok to use any way to pass information. In this case, the library chose to use exceptions to this end, so we are ok to catch them. This is a little ugly, but it's a lot less ugly than using a map as a set.

If you are going to leave the map, please at least add a comment to the field to explain that the map is actually used to model a set and why it is done so.

**Review by rpuch — APPROVED** [review]

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java:49]: You are right, I should write about it in the comments.

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:632]: Cool, I didn't realize that. Poor naming, in my opinion

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java [hunk 25]
```diff
@@ -604,148 +605,148 @@ void testScanSimple() {
...
-                                null,
-                                (GREATER | LESS)
-                        ).stream()
-                        .map(indexRow -> serializer.deserializeColumns(indexRow)[0])
-                        .collect(toList()),
+                scan(
+                        indexStorage,
+                        serializer.serializeRowPrefix(1),
+                        null,
+                        (GREATER | LESS),
```
Comment (ibessonov) [inline modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:696]: Just a question, why did you add parentheses here?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java [hunk 26]
```diff
@@ -171,14 +172,15 @@ private Cursor<ByteBuffer> createScanCursor(byte @Nullable [] lowerBound, byte @
...
-            private byte[] key;
+            private byte @Nullable [] key;
 
             @Override
             public void close() {
-                it.close();
-
-                RocksUtils.closeAll(options, upperBoundSlice);
+                try {
+                    closeAll(it, () -> RocksUtils.closeAll(options, upperBoundSlice));
```
Comment (ibessonov) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:180]: Wait a minute ,why not `closeAll(it, options, upperBoundSlice)`? What's the reason for creating a lambda?

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java (hunk 25 above)
Comment (tkalkirill) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java:696]: For clarity

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java (hunk 26 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java:180]: Fix it

**Review by ibessonov — APPROVED** [review]
